### ISSUE TITLE
gnome: improve compile_resources

### ISSUE THREAD

**Issue by matthiasclasen** [mesonbuild/meson#9334]
The way compile_resources works is high-overhead for big resource bundles: It first generates the bundle, then it serializes it into a C string, then it feeds it to gcc. When compiling GTK, this causes a noticable slowdown for applications that have big resource bundles.

glib-compile-resources supports an --external-data option which avoids the serialization, and instead leaves it up to the user to produce a .o file with the data in it. On linux, this can be done with ld and objcopy. The necessary incantations can be seen here:

https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4019/diffs?commit_id=6e245f64788b99cac7e5fa9172b95bf017474de7

It would be fantastic if the gnome module did this for us.

**Comment by eli-schwartz** [comment]
More generally, it should not have to be gnome or qt's job to create resource bundles in the first place... because in a sane world we'd have http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2592.htm / https://thephd.dev/embed-the-details

The `ld` trick is described there and not in a flattering way. Then again, apparently nothing works well and ~~the entire universe is a disaster~~. Anyway, this is apparently not particularly portable.

**Comment by eli-schwartz** [comment]
It seems your linked MR is closed without being merged?

**Comment by matthiasclasen** [comment]
yeah, but the commits ended up on master anyway.

**Comment by matthiasclasen** [comment]
> The ld trick is described there and not in a flattering way.

I don't think that complaint is relevant for out usecases. We generally don't want the data to be optimized away after we worked hard to include it.


> The ld trick is described there and not in a flattering way. Then again, apparently nothing works well and the entire universe is a disaster. Anyway, this is apparently not particularly portable.

The portable part is what we do now, and it gets very slow. I'm asking to do the nonportable, fast thing where it works.

**Comment by smcv** [comment]
> glib-compile-resources supports an --external-data option which avoids the serialization, and instead leaves it up to the user to produce a .o file with the data in it. On linux, this can be done with ld and objcopy

Note that this seems to be tricky to get right: https://gitlab.gnome.org/GNOME/gtk/-/issues/4598, https://gitlab.gnome.org/GNOME/gtk/-/issues/4617.
